scipy/stats/_hypotests.py

```diff
@@ -9,6 +9,276 @@
+The statistic of each axis-slice (e.g. row) of the input will appear in a
+corresponding element of the output.
+If ``None``, the input will be raveled before computing the statistic."""
+.split('\n'))
+_axis_parameter_doc = Parameter(_name, _type, _desc)
+_axis_parameter = inspect.Parameter(_name,
+                                    inspect.Parameter.KEYWORD_ONLY,
+                                    default=0)
+
+_name = 'nan_policy'
+_type = "{'propagate', 'omit', 'raise'}"
+_desc = (
+"""Defines how to handle input NaNs.
+
+- ``propagate``: if a NaN is present in the axis slice (e.g. row) along
+  which the  statistic is computed, the corresponding entry of the output
+  will be NaN.
+- ``omit``: NaNs will be omitted when performing the calculation.
+  If insufficient data remains in the axis slice along which the
+  statistic is computed, the corresponding entry of the output will be
+  NaN.
+- ``raise``: if a NaN is present, a ``ValueError`` will be raised."""
+.split('\n'))
+_nan_policy_parameter_doc = Parameter(_name, _type, _desc)
+_nan_policy_parameter = inspect.Parameter(_name,
+                                          inspect.Parameter.KEYWORD_ONLY,
+                                          default='propagate')
+
+
+def _vectorize_hypotest_factory(result_object, default_axis=0,
+                                n_samples=1, paired=False,
+                                result_unpacker=None, too_small=0,
+                                nan_policy_position=None, axis_position=None):
+    """Factory for a wrapper that adds axis/nan_policy params to a function.
+
+    Parameters
+    ----------
+    result_object : callable
+        Callable that returns an object of the type returned by the function
+        being wrapped (e.g. the namedtuple or dataclass returned by a
+        statistical test) provided the separate components (e.g. statistic,
+        pvalue).
+    default_axis : int, default: 0
+        The default value of the axis argument. Standard is 0 except when
+        backwards compatibility demands otherwise (e.g. `None`).
+    n_samples : int, default: 1
+        The number of data samples accepted by the function. For example,
+        `gmean` accepts one sample, `ks_2samp` accepts two samples. `None`
+        means an arbitrary number of samples (e.g. `fligner`).
+    paired : {False, True}
+        Whether the function being wrapped treats the samples as paired (i.e.
+        corresponding elements of each sample should be considered as different
+        components of the same sample.)
+    result_unpacker : callable, optional
+        Function that unpacks the results of the function being wrapped into
+        a tuple. This is essentially the inverse of `result_object`. Default
+        is `None`, which is appropriate for statistical tests that return a
+        statistic, pvalue tuple (rather than, e.g., a non-iterable datalass).
+    too_small : int, default: 0
```

Comment:
> It would almost call for a decorator to do input validation instead of what we currently do

While some input validation can be standardized, that is not the intent here. Its purpose is to address [this](https://github.com/scipy/scipy/pull/13312#issuecomment-837975864).

Reply:
Sure sure, I was just making an observation here.